services/QuillLMS/app/services/user_pack_sequence_item_saver.rb

```diff
@@ -0,0 +1,52 @@
+# frozen_string_literal: true
+
+class UserPackSequenceItemSaver < ApplicationService
+  attr_reader :classroom_id, :user_id, :user_pack_sequence_item_statuses
+
+  LOCKED = UserPackSequenceItem::LOCKED
+  UNLOCKED = UserPackSequenceItem::UNLOCKED
+  COMPLETED_KEY = UserPackSequenceItemQuery::COMPLETED_KEY
+  PACK_SEQUENCE_ITEM_ID_KEY = UserPackSequenceItemQuery::PACK_SEQUENCE_ITEM_ID_KEY
+
+  def initialize(classroom_id, user_id)
+    @classroom_id = classroom_id
+    @user_id = user_id
+    @user_pack_sequence_item_statuses = {}
+  end
+
+  def run
+    compute_statuses
```

Comment:
I'm curious why you decided to do this in two passes, compute everything THEN store everything vs. compute and store in one loop?

Asking mainly because I think you want the `select`s and the `write`s both in the `transaction` to prevent race conditions (e.g. the data changes between when you read it and when you write to the DB).

Reply:
My thought for breaking this up was to limit the number of update!(status: status) to one per pack_sequence_item.  Otherwise, you end up doing an update per unit_activity in the pack_sequence_item.

However, you raise a good point that this leaves us open to race conditions, so I would agree on putting them inside the transaction instead and doing the redundant updates.